### PR TITLE
Add query proof

### PR DESCRIPTION
**Pull request by greatect**
Resolving part of issue #25:
* Use IAVL tree to maintain state and provide merkle proof
* Deal with app height, hash, epoch and put them into ABCI info
* Provide proofs inside ABCI query responses
* Add a test for posting document and querying documents with proof
* Add utilities for generating documents, but still not auto generated during the commit stage
* Remove duplicate signing/verification of document
* Refinement on storage keys
* Naming on error codes

### CODE REVIEW

File: util.go [hunk 1]
```diff
@@ -0,0 +1,219 @@
...
+	key = make([]byte, len(keyPrefix))
+	copy(key, keyPrefix)
+	key = append(key[:], ':')
+	key = append(key[:], verHex[:]...)
+	key = append(key[:], ':')
+	key = append(key[:], IDHex[:]...)
+	return
+}
+
+func (state *KatzenmintState) documentForEpoch(epoch uint64) ([]byte, merkle.ProofOperator, error) {
```
Comment (sc0Vu) [inline util.go:40]: IMHO we should put this function into state.go, If we only use this function in katzenmint state.

File: util.go [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+		if epoch > now+1 {
+			return nil, nil, fmt.Errorf("requesting document for a too future epoch %d", epoch)
+		}
+		return nil, nil, fmt.Errorf("document for epoch %d is not ready yet", epoch)
+	}
+	valueOp := iavl.NewValueOp(key, proof)
+	return doc, valueOp, nil
+}
+
+func (s *KatzenmintState) generateDocument() (*document, error) {
```
Comment (sc0Vu) [inline util.go:69]: Same.

**Review by sc0Vu — CHANGES_REQUESTED** [review]

File: helper.go [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+	"github.com/hashcloak/katzenmint-pki/s11n"
+	"github.com/katzenpost/core/crypto/ecdh"
+	"github.com/katzenpost/core/crypto/eddsa"
+	"github.com/katzenpost/core/crypto/rand"
+	"github.com/katzenpost/core/pki"
+	"github.com/stretchr/testify/require"
+)
+
+// create test descriptor
+func CreateTestDescriptor(require *require.Assertions, idx int, layer int, epoch uint64) (*pki.MixDescriptor, []byte) {
```
[the remark this answers is not in the document]
Comment (greatect) [reply helper.go]: Moved to a separate testutil package

File: helper.go [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+	require.NoError(err, "IsDescriptorWellFormed(good)")
+
+	// Sign the descriptor.
+	signed, err := s11n.SignDescriptor(identityPriv, desc)
+	require.NoError(err, "SignDescriptor()")
+	return desc, signed
+}
+
+// create test document
+func CreateTestDocument(require *require.Assertions, epoch uint64) (*s11n.Document, []byte) {
```
[the remark this answers is not in the document]
Comment (greatect) [reply helper.go]: Moved to a separate testutil package

**Review by sc0Vu — APPROVED** [review]
LGTM

**Pull request merged**